Fixes #244, #154, #196, #173, #235 and #239

alganet opened the pull request:
- Moves code from  __invoke on AbstractRule into NotEmpty.
  - Moves foreign type check to the chain.
  - Fixes tests using the chain instead of concrete instances.
  - Adds a parameter for strict type checking on NotEmpty. (no BC break)
  - Adds new tests for the BC behavior and new parameter.
  - Fixes AbstractRelated to allow empty referenced values.
  - Removes dead code found in #239 

Fixes #154, #196, #173 and #235.

henriquemoody commented on README.md:
```diff
@@ -1475,14 +1475,19 @@ See also:
   * [v::noneOf()](#vnoneofv1-v2-v3)
 
-#### v::notEmpty()
+#### v::notEmpty($jugglesTypes = true)
```
You must update the links too [links](https://github.com/Respect/Validation#vnotempty)

henriquemoody commented:
@alganet, since you're doing this, cal you fix #235 too?

henriquemoody commented on README.md:
```diff
@@ -1475,14 +1475,19 @@ See also:
+v::string()->notEmpty()->validate('0'); //false
+
+// Disable type juggling
+v::string()->notEmpty(false)->validate('0'); //true
```
I was expecting `v::int()->notEmpty(false)->validate(0)` to return `true` too, but it didn't happen

alganet replied:
Do you think it should? That would make `notEmpty` a string-only rule.

henriquemoody replied:
My expectations in non-strict mode:

| Value             | NotEmpty |
| -------------     | ---------|
| `null`            | No       |
| `""`              | No       |
| `array()`         | Yes      |
| `"0"`             | Yes      |
| `0`               | Yes      |
| `"0.0"`           | Yes      |
| `0.0`             | Yes      |
| `true`            | Yes      |

Does it makes sense for you? Take a look on #205.

henriquemoody commented on library/Rules/NotEmpty.php:
```diff
@@ -3,12 +3,31 @@
+
+    public function __invoke($input)
+    {
+        return $this->validate($input);
```
What you did here and on `AbstractRule` was great! Very clever :clap:

henriquemoody commented on library/Rules/NotEmpty.php:
```diff
@@ -3,12 +3,31 @@
-            $input = trim($input);
+        if (!is_string($input)) {
+            return !empty($input);
+        }
```
Just an idea for the first major version: when the input is an array we can use `array_filter()` to remove empty values.

alganet commented:
I've added the fix for #235. It was actually a bug on `AbstractRelated` (the base class for Key, Attribute, Each and indirect validators).

Guys, I'm not sure about the parameter on `NotEmpty`. I've changed it to a simpler name. My implementation only disallows quoted zeroes like `'  0'` and `"0"` (with or without whitespace), which seems to be the cause of the trouble.

henriquemoody commented:
Don't forget to update Validator docblock, too

henriquemoody commented on README.md:
```diff
@@ -1476,13 +1476,19 @@ See also:
   * [v::noneOf()](#vnoneofv1-v2-v3)
 
 #### v::notEmpty()
+#### v::notEmpty($strict = false)
```
Not a blocker, but I guess `v::notEmpty(boolean $strict)` is better.

henriquemoody commented:
#235 is not fixed yet, I think to fix that we have to overwrite all possible `validate()` methods...

malukenho commented on library/Rules/AbstractRelated.php:
```diff
@@ -66,10 +66,12 @@ public function check($input)
 
-        return $this->decision('validate', $hasReference, $input);
+        return $this->decision('validate', $hasReference, $input) ||
+        $this->getReferenceValue($input) === '';
```
`empty()`

henriquemoody commented:
I will close this pull request because `NotEmpty` has a new feature and new features must be send against to `master` branch.
Also I toke some part of code on this pull request and I sent #298.

About the NotEmpty, I was wondering about this, I'm in favor of creating a new rule, called `NotBlank` which should have more or less the same behavior of our current `NotEmpty` and the `NotEmpty` rule should use `empty()` but since it's a major BC it's better we do this in the 1.0 version, IMO.

Any way, thanks you very much for this pull request, #298 is already available on `0.7.5` and `0.8.1` versions and on `master` branch as well.
